src/extension/prelude.rs

```diff
@@ -8,13 +8,36 @@ use crate::{
     ops::LeafOp,
     types::{
         type_param::{TypeArg, TypeParam},
-        CustomCheckFailure, CustomType, FunctionType, Type, TypeBound,
+        CustomCheckFailure, CustomType, FunctionType, PolyFuncType, Type, TypeBound,
     },
     values::{CustomConst, KnownTypeConst},
     Extension,
 };
 
-use super::ExtensionRegistry;
+use super::{ExtensionRegistry, SignatureError, SignatureFromArgs};
+struct ArrayOpCustom;
+
+const MAX: &[TypeParam; 1] = &[TypeParam::max_nat()];
+impl SignatureFromArgs for ArrayOpCustom {
+    fn compute_signature(&self, arg_values: &[TypeArg]) -> Result<PolyFuncType, SignatureError> {
+        let [TypeArg::BoundedNat { n }] = *arg_values else {
+            panic!("Should have been checked already.")
+        };
+        let elem_ty_var = Type::new_var_use(0, TypeBound::Any);
+
+        let var_arg_row = vec![elem_ty_var.clone(); n as usize];
+        let other_row = vec![array_type(elem_ty_var.clone(), TypeArg::BoundedNat { n })];
+
+        Ok(PolyFuncType::new(
+            vec![TypeParam::Type(TypeBound::Any)],
```

Comment:
Are we allowing arrays of linear types?

Reply:
it was the case before - I haven't changed anything